### DISCUSSION TITLE
обучение двум видам оружия

### DISCUSSION THREAD

**Discussion by Kor-Angar** [dosinabox/g2_classic#89] [Ideas]
в готики 2 классик, происходит тренировка обоим видам оружия. в диалоге это не отображено, из-за чего происходит путаница

**Comment by Kor-Angar** [comment]
тестово попробовал сочинить вот такое название. есть предложения или замечания?
![20220807164607_1](https://user-images.githubusercontent.com/49514764/183293971-2a6f8b0f-e486-4cae-aa64-478f380ed16d.jpg)

**Reply by Kor-Angar** [reply]
```D
    PRINT_LearnAllHanded1 = Str_Format("[%s & %s] +1", NAME_OneHanded, NAME_TwoHanded);
    PRINT_LearnAllHanded5 = Str_Format("[%s & %s] +5", NAME_OneHanded, NAME_TwoHanded);
```

**Comment by Kor-Angar** [comment]
![20220809102151_1](https://user-images.githubusercontent.com/49514764/183589227-fd394f29-cf28-4af7-b7fb-e408fa58a004.jpg)

**Reply by Kor-Angar** [reply]
```D
var string PRINT_LearnAllHanded1;
var string PRINT_LearnAllHanded5; 

namespace zLearnAllHanded 
{
    func event GameInit() 
    {
        PRINT_LearnAllHanded1 = Str_Format("[%s & %s] +1", NAME_OneHanded, NAME_TwoHanded);
        PRINT_LearnAllHanded5 = Str_Format("[%s & %s] +5", NAME_OneHanded, NAME_TwoHanded); 
    };
};


```

**Reply by dosinabox** [reply]
Одноручное и двуручное оружие + x (y очков обучения)

**Reply by dosinabox** [reply]
Если хочется с константами, то так ок. Только квадратные скобки лишние.
Одноручное оружие & Двуручное оружие + x (y очков обучения)

**Reply by dosinabox** [reply]
```cpp
"%s & %s + 1", NAME_OneHanded, NAME_TwoHanded);
"%s & %s + 5", NAME_OneHanded, NAME_TwoHanded);
```

**Reply by dosinabox** [reply]
> Одноручное и двуручное оружие + x (y очков обучения)

Я может так себе добавлю. Подскажешь куда?

**Reply by Kor-Angar** [reply]
Мне кажется что без скобок неочевидно будет что оба навка поднимется. но в целом это уже букваедство, так как общий смысл и так будет понятен что тренироваться будет два навыка и что это не "штраф" :-)

**Comment by Kor-Angar** [comment]
Конечно. Тестово ставил на два учителя. К компу вернусь скину два варианта.

**Reply by Kor-Angar** [reply]
@dosinabox тестил на Корде и Вульфгаре.

```D
var string 	DIA_CurrentName;
var int 	DIA_CurrentInstance;
const int 	DIA_CurrentCounter = 0; 

var string PRINT_LearnAllHanded1;
var string PRINT_LearnAllHanded5; 

namespace zLearnAllHanded 
{
    func event GameInit() 
    {
        PRINT_LearnAllHanded1 = Str_Format("[%s & %s] +1", NAME_OneHanded, NAME_TwoHanded);
        PRINT_LearnAllHanded5 = Str_Format("[%s & %s] +5", NAME_OneHanded, NAME_TwoHanded); 
    };
};
```

[DIA_MIL_312_Wulfgar.zip](https://github.com/dosinabox/g2_classic/files/9324721/DIA_MIL_312_Wulfgar.zip)
[DIA_SLD_805_Cord.zip](https://github.com/dosinabox/g2_classic/files/9324722/DIA_SLD_805_Cord.zip)

**Reply by Kor-Angar** [reply]
`var int 	DIA_CurrentInstance;` сделал Гратт для zparser 
практически это позволит использовать `Info_AddChoice` в разных инстанциях.

**Reply by Kor-Angar** [reply]
это про `DIA_CurrentInstance` там ничего нового ты не увидишь, но продублирую на всякий случай
[message.zip](https://github.com/dosinabox/g2_classic/files/9324752/message.zip)

**Reply by Kor-Angar** [reply]
`var string 	DIA_CurrentName;` для отладчика

**Reply by Kor-Angar** [reply]
`const int 	DIA_CurrentCounter = 0; ` это счётчик внутри инстанции. он тут не нужен.

**Reply by Kor-Angar** [reply]
> const int DIA_CurrentCounter = 0;

 это счётчик внутри инстанции. он тут не нужен. обнуляется при каждом запуске функции инфо

**Comment by dosinabox** [comment]
А почему просто не добавить условие в b_buildlearnstring, которую используют все учителя? Проверь так:
```cpp
const string PRINT_LEARN1H2H1 = "Одноручное и двуручное оружие + 1";
const string PRINT_LEARN1H2H5 = "Одноручное и двуручное оружие + 5";
const string PRINT_LEARN1H2H10 = "Одноручное и двуручное оружие + 10";
const string PRINT_LEARN1H2H20 = "Одноручное и двуручное оружие + 20";
const string PRINT_LEARNBOWCROSSBOW1 = "Лук и арбалет + 1";
const string PRINT_LEARNBOWCROSSBOW5 = "Лук и арбалет + 5";
const string PRINT_LEARNBOWCROSSBOW10 = "Лук и арбалет + 10";
const string PRINT_LEARNBOWCROSSBOW20 = "Лук и арбалет + 20";

func string b_buildlearnstring(var string text,var int kosten)
{
	var string concattext;
	if(Npc_GetTalentSkill(other,NPC_TALENT_1H) > Npc_GetTalentSkill(other,NPC_TALENT_2H))
	{
		if(Hlp_StrCmp(text,PRINT_LEARN1H1))
		{
			text = PRINT_LEARN1H2H1;
		}
		else if(Hlp_StrCmp(text,PRINT_LEARN1H5))
		{
			text = PRINT_LEARN1H2H5;
		}
		else if(Hlp_StrCmp(text,PRINT_LEARN1H10))
		{
			text = PRINT_LEARN1H2H10;
		}
		else if(Hlp_StrCmp(text,PRINT_LEARN1H20))
		{
			text = PRINT_LEARN1H2H20;
		};
	}
	else if(Npc_GetTalentSkill(other,NPC_TALENT_2H) > Npc_GetTalentSkill(other,NPC_TALENT_1H))
	{
		if(Hlp_StrCmp(text,PRINT_LEARN2H1))
		{
			text = PRINT_LEARN1H2H1;
		}
		else if(Hlp_StrCmp(text,PRINT_LEARN2H5))
		{
			text = PRINT_LEARN1H2H5;
		}
		else if(Hlp_StrCmp(text,PRINT_LEARN2H10))
		{
			text = PRINT_LEARN1H2H10;
		}
		else if(Hlp_StrCmp(text,PRINT_LEARN2H20))
		{
			text = PRINT_LEARN1H2H20;
		};
	};
	if(Npc_GetTalentSkill(other,NPC_TALENT_BOW) > Npc_GetTalentSkill(other,NPC_TALENT_CROSSBOW))
	{
		if(Hlp_StrCmp(text,PRINT_LEARNBOW1))
		{
			text = PRINT_LEARNBOWCROSSBOW1;
		}
		else if(Hlp_StrCmp(text,PRINT_LEARNBOW5))
		{
			text = PRINT_LEARNBOWCROSSBOW5;
		}
		else if(Hlp_StrCmp(text,PRINT_LEARNBOW10))
		{
			text = PRINT_LEARNBOWCROSSBOW10;
		}
		else if(Hlp_StrCmp(text,PRINT_LEARNBOW20))
		{
			text = PRINT_LEARNBOWCROSSBOW20;
		};
	}
	else if(Npc_GetTalentSkill(other,NPC_TALENT_CROSSBOW) > Npc_GetTalentSkill(other,NPC_TALENT_BOW))
	{
		if(Hlp_StrCmp(text,PRINT_LEARNCROSSBOW1))
		{
			text = PRINT_LEARNBOWCROSSBOW1;
		}
		else if(Hlp_StrCmp(text,PRINT_LEARNCROSSBOW5))
		{
			text = PRINT_LEARNBOWCROSSBOW5;
		}
		else if(Hlp_StrCmp(text,PRINT_LEARNCROSSBOW10))
		{
			text = PRINT_LEARNBOWCROSSBOW10;
		}
		else if(Hlp_StrCmp(text,PRINT_LEARNCROSSBOW20))
		{
			text = PRINT_LEARNBOWCROSSBOW20;
		};
	};
	concattext = ConcatStrings(text,PRINT_KOSTEN);
	concattext = ConcatStrings(concattext,IntToString(kosten));
	if(kosten >= 5)
	{
		return ConcatStrings(concattext,PRINT_LP);
	};
	if(kosten > 1)
	{
		return ConcatStrings(concattext,PRINT_2LP);
	};
	return ConcatStrings(concattext,PRINT_1LP);
};
```

**Reply by Kor-Angar** [reply]
> А почему просто не добавить условие в b_buildlearnstring, которую используют все учителя?

а я всё время забываю про сравнение строчек `Hlp_StrCmp` так как их не использую :-)

> Проверь так

попробую

**Reply by Kor-Angar** [reply]
работает.

**Reply by dosinabox** [reply]
Тогда оставляю. Надо будет релиз сделать скоро, накопилось дополнений.

**Reply by Kor-Angar** [reply]
а  ты `Str_Format` еще не начала использовать? :-)

**Reply by dosinabox** [reply]
> а ты `Str_Format` еще не начала использовать? :-)

Пока не доводилось)

**Reply by Kor-Angar** [reply]
> Тогда оставляю.

а еще луки и арбалеты ;-)

**Reply by Kor-Angar** [reply]
сначала не привычно. как писал Пушкин , привычка свыше нам дана, замена счастью есть она :-)
```D
	//concattext = ConcatStrings(text,PRINT_KOSTEN);
	//concattext = ConcatStrings(concattext,IntToString(kosten));
    concattext = Str_Format("%s%s%i", text, PRINT_KOSTEN,kosten);
```
я вот всё хочу статистику  переписать с `Str_Format` но мне страшно открыть код :-)

**Reply by Kor-Angar** [reply]
> привычка свыше нам дана, замена счастью есть она 

опа. я даже не перепутал ни одного слова :-)

**Reply by Kor-Angar** [reply]
ан нет. слово "есть" лишнее. **Замена счастию она**

**Reply by Kor-Angar** [reply]
> Надо будет релиз сделать скоро

заклинания еще дотестить надо.

**Reply by dosinabox** [reply]
> а еще луки и арбалеты ;-)

Добавлены в код выше, проверь

**Reply by Kor-Angar** [reply]
работает

**Comment by Kor-Angar** [comment]
глюк какой то. надо будет перепроверить из-за чего
![20230223042008_1](https://user-images.githubusercontent.com/49514764/220812575-9108125e-9032-48f8-b5cc-634de21e909c.jpg)

**Reply by Kor-Angar** [reply]
![20230223031030_1](https://user-images.githubusercontent.com/49514764/220812826-a4703130-c3f7-4d1a-b68b-060384353e4d.jpg)
